### ISSUE TITLE
changelog : `llama-server` REST API

### ISSUE THREAD

**Issue by ggerganov** [ggml-org/llama.cpp#9291]
# Overview

This is a list of changes to the public HTTP interface of the `llama-server` example. Collaborators are encouraged to edit this post in order to reflect important changes to the API that end up merged into the `master` branch.

If you are building a 3rd party project that relies on `llama-server`, it is recommended to follow this issue and check it carefully before upgrading to new versions.

See also:

- [Changelog for `libllama` API](https://github.com/ggerganov/llama.cpp/issues/9289)

## Recent API changes (most recent at the top)

| version | PR  | desc |
| ---     | --- | ---  |
| b4599 | #9639 | `/v1/chat/completions` now supports `tools` & `tool_choice` |
| TBD.  | #10974 | `/v1/completions` is now OAI-compat |
| TBD.  | #10783 | `logprobs` is now OAI-compat, default to pre-sampling probs |
| TBD.  | #10861 | `/embeddings` supports pooling type `none` |
| TBD.  | #10853 | Add optional `"tokens"` output to `/completions` endpoint |
| b4337 | #10803 | Remove `penalize_nl` |
| b4265 | #10626 | CPU docker images working directory changed to /app |
| b4285 | #10691 | (Again) Change `/slots` and `/props` responses |
| b4283 | #10704 | Change `/slots` and `/props` responses |
| b4027 | #10162 | `/slots` endpoint: remove `slot[i].state`, add `slot[i].is_processing` |
| b3912 | #9865 | Add option to time limit the generation phase |
| b3911 | #9860 | Remove self-extend support |
| b3910 | #9857 | Remove legacy system prompt support |
| b3897 | #9776 | Change default security settings, `/slots` is now disabled by default<br/>Endpoints now check for API key if it's set |
| b3887 | #9510 | Add `/rerank` endpoint |
| b3754 | #9459 | Add `[DONE]\n\n` in OAI stream response to match spec |
| b3721 | #9398 | Add `seed_cur` to completion response |
| b3683 | #9308 | Environment variable updated |
| b3599 | #9056 | Change `/health` and `/slots` |

*For older changes, use:*

```bash
git log --oneline -p b3599 -- examples/server/README.md
```

## Upcoming API changes

- TBD

**Comment by ngxson** [comment]
Not a REST API breaking change, but is server-related: some environment variables are changed in https://github.com/ggerganov/llama.cpp/pull/9308

**Comment by slaren** [comment]
After #9398, in the completion response `seed` contains the seed requested by the user, while `seed_cur` contains the seed used to generate the completion. The values can be different if `seed` is `LLAMA_DEFAULT_SEED` (or -1), in which case a random seed is generated and returned in `seed_cur`.

**Comment by isaac-mcfadyen** [comment]
> Breaking change for `/slots` endpoint #10162
> 
> `slot[i].state` is removed and replaced by `slot[i].is_processing`
> 
> `slot[i].is_processing === false` means the slot is idle

Was the `slots` endpoint also disabled by default? (or maybe just a documentation change?)
https://github.com/ggerganov/llama.cpp/pull/10162/files#diff-42ce5869652f266b01a5b5bc95f4d945db304ce54545e2d0c017886a7f1cee1aR698

**Comment by ngxson** [comment]
For security reasons, "/slots" was disabled by default since https://github.com/ggerganov/llama.cpp/pull/9776 , and was mentioned in the breaking changes table. I just forgot to update the docs.

**Comment by ngxson** [comment]
Not an API change, but maybe good to know that the default web UI for `llama-server` changed in https://github.com/ggerganov/llama.cpp/pull/10175

If you want to use the old completion UI, please follow instruction in the PR.

**Comment by ggerganov** [comment]
`cache_prompt: true` is now used by default (#10501)

**Comment by ngxson** [comment]
`/props/`and `/slots` endpoints has changed in #10691 and #10704 , see server/README.md for more

**Comment by ngxson** [comment]
`/embeddings` will NOT be OAI-compat after #10861 

For clarification, we will maintain OAI-compat for all API under `/v1` prefix, including:
- `/v1/embeddings`
- `/v1/chat/completions`

NOTE: OAI support for `/v1/completions` will come in the near future

**Comment by ngxson** [comment]
Behavior of `n_probs` has changed in https://github.com/ggerganov/llama.cpp/pull/10783 , we're now providing OAI-compatible `logprobs` option

**Comment by ngxson** [comment]
Added OAI-compat support for `/v1/completions` here: https://github.com/ggerganov/llama.cpp/pull/10974

If you want to use it with downstream library, be sure to add `/v1` prefix. For example, using python library:

```py
from openai import OpenAI

client = OpenAI(api_key="dummy", base_url=f"http://localhost:8080/v1")
res = client.completions.create(
    model="davinci-002",
    prompt="I believe the meaning of life is",
    max_tokens=8,
)
```

If you want to use the old non-OAI style, remove the `/v1` from endpoint path.

**Comment by isaac-mcfadyen** [comment]
Guessing we want to keep this open.

**Comment by aviallon** [comment]
We need to add the new vision support

**Comment by Ifiht** [comment]
Along with the vision support, are there currently plans to add the MCP tool-calling function? Or maybe it's already here & I just don't know where to look?
[OpenAI MCP Ref](https://platform.openai.com/docs/guides/tools-remote-mcp)

**Comment by 65a** [comment]
Wasn't sure how to add to the API changelog...

#15108 is a forward compatible API change to support multimodal in the embeddings and completion endpoints. It also adds the capability "multimodal" to the models endpoint when llama-server was started with a multimodal-capable model and mmproj.

**Comment by ngxson** [comment]
#15827 : added `return_progress` and `timings.cache_n`, useful for tracking prompt processing progress and context length usage

**Comment by Cyrille37** [comment]
Hi. I hope I'm in the right place for talking about #9639 :-)

In [server/server-chat.cpp#](https://github.com/ggml-org/llama.cpp/blob/master/tools/server/server-chat.cpp#L491) the `tool_choice` is forced to String as "auto" or "required". That generates a log "_Wrong type supplied for parameter 'tool_choice'. Expected 'string', using default value: [json.exception.type_error.302] type must be string, but is object_".

- Is there a reason for this forced type casting ? Perhaps a standby while the full OpenAI specification will be implemented ?
- Would the log be replaced from "Wrong" to "Notice" or "Info" ?

Regards & cheers.

**Comment by fatinm1** [comment]
A few things I noticed that aren't reflected in the table or the "Upcoming API changes" section yet:

    -    Upcoming: #26508 added a runtime warning that the default server port will change from :8080 to :9931 in a future release —     seems worth listing under "Upcoming API changes" rather than leaving it as TBD.
    -    #26347 made the /models and /props endpoints require the API key when authentication is enabled (previously accessible regardless).
    -    #26045 added support for "reasoning_effort": "none" in the OAI-compatible chat API.
